core/executor/src/allocator.rs

```diff
@@ -382,7 +445,43 @@ mod tests {
 		if let Err(err) = ptr {
 			match err {
 				Error::RequestedAllocationTooLarge => {},
-				e => panic!("Expected out of space error, got: {:?}", e),
+				e => panic!("Expected allocation size too large error, got: {:?}", e),
+			}
+		}
+	}
+
+	#[test]
+	fn should_return_error_when_bumper_greater_than_heap_size() {
+		// given
+		let mem = MemoryInstance::alloc(Pages(1), None).unwrap();
+		let mut heap = FreeingBumpHeapAllocator::new(mem, 0);
+		heap.max_heap_size = 64;
+
+		let ptr1 = heap.allocate(32).unwrap();
+		assert_eq!(ptr1, PREFIX_SIZE);
+		heap.deallocate(ptr1).expect("failed freeing ptr1");
+		assert_eq!(heap.total_size, 0);
+		assert_eq!(heap.bumper, 40);
+
+		let ptr2 = heap.allocate(16).unwrap();
+		assert_eq!(ptr2, 48);
+		heap.deallocate(ptr2).expect("failed freeing ptr2");
+		assert_eq!(heap.total_size, 0);
+		assert_eq!(heap.bumper, 64);
+
+		// when
+		// the `bumper` value is equal to `max_heap_size` here and any
+		// further allocation which would increment the bumper must fail.
+		// we try to allocate 8 bytes here, which will increment the
+		// bumper since no 8 byte item has been allocated+freed before.
+		let ptr = heap.allocate(8);
+
+		// then
+		assert_eq!(ptr.is_err(), true);
```

Comment:
nit: `assert!(ptr.is_err())`

Or just match directly on `ptr` and panic on `Ok`.